blocks/bricks/recurrent.py

```diff
@@ -300,6 +301,116 @@ def apply(self, input_=None, state=None, mask=None):
+    This implementation aims to do as many computations in parallel as
+    possible and expects the last dimension of the input to be four times
+    the output dimension.
+
+    Parameters
+    ----------
+    dim : int
+        The dimension of the hidden state.
+
+    Notes
+    -----
+    See :class:`.Initializable` for initialization parameters.
+
+    """
+    @lazy
+    def __init__(self, dim, **kwargs):
+        super(LSTM, self).__init__(**kwargs)
+        self.dim = dim
+
+    def get_dim(self, name):
+        if name == 'inputs':
+            return self.dim * 4
+        else:
+            return self.dim
+
+    def _allocate(self):
+        self.W_state = shared_floatx_zeros((self.dim, 4*self.dim),
+                                           name='W_state')
+        self.W_cell = shared_floatx_zeros((self.dim, 2*self.dim),
+                                          name='W_cell')
+        self.W_cell_to_gate = shared_floatx_zeros((self.dim, self.dim),
+                                                  name='W_cell_to_gate')
+        self.biases = shared_floatx_zeros((4*self.dim,), name='biases')
+        add_role(self.W_state, WEIGHTS)
+        add_role(self.W_cell, WEIGHTS)
+        add_role(self.W_cell_to_gate, WEIGHTS)
+        add_role(self.biases, BIASES)
+
+        self.params = [self.W_state, self.W_cell, self.W_cell_to_gate,
+                       self.biases]
+
+    def _initialize(self):
+        self.biases_init.initialize(self.biases, self.rng)
+        for w in self.params[:-1]:
+            self.weights_init.initialize(w, self.rng)
+
+    @recurrent(sequences=['inputs', 'mask'], states=['states', 'cells'],
+               contexts=[], outputs=['states', 'cells'])
+    def apply(self, inputs, states, cells, mask=None):
+        """Apply the Long Short Term Memory transition.
+
+        Parameters
+        ----------
+        states : :class:`~tensor.TensorVariable`
+            The 2 dimensional matrix of current states in the shape
+            (batch_size, features). Required for `one_step` usage.
+        cells : :class:`~tensor.TensorVariable`
+            The 2 dimensional matrix of current cells in the shape
+            (batch_size, features). Required for `one_step` usage.
+        inputs : :class:`~tensor.TensorVariable`
+            The 2 dimensional matrix of inputs in the shape (batch_size,
+            features * 4).
+        mask : :class:`~tensor.TensorVariable`
+            A 1D binary array in the shape (batch,) which is 1 if there is
+            data available, 0 if not. Assumed to be 1-s only if not given.
+
+        Returns
+        -------
+        states : :class:`~tensor.TensorVariable`
+            Next states of the network.
+        cells : :class:`~tensor.TensorVariable`
+            Next cell activations of the network.
+
+        """
+        # Unfortunately, I'm not aware of a general way to slice the last
+        # dimension of a theano tensor.
```

Comment:
I fixed it with the transpose for now. Not sure why I didn't think of that myself :)